### PR TITLE
tracing and profiling interface and basic backends

### PR DESCRIPTION
**Pull request by trws**
note: this is a sanitized branch, commit cleanup is the main difference

This is a *very* rough cut on tracing for flux.  If TAU_PROFILING is
enabled, it turns on some macros that will register flux event
processing with TAU.  Getting flux to build, link, and run with TAU is
another matter entirely.   I recommend building your own TAU with only
pthread support, then specifying tau_cc.sh as the compiler with only
the options to use shared linking and track pthreads while only
affecting linking.  Then when running preload the pthread indirection
library (this should not be necessary, but it is).

Alternately, there is a caliper backend to the basic macros as well,
which can be turned on with the `--with-caliper` option on configure.
One still has to have caliper available in appropriate paths, but it's
much easier to get started with this.  Unfortunately the data is rather
harder to wrangle from this right now.

Either way, the main point of this is to serve as a starting point for
tracing/profiling in flux rather than a full setup.

### CODE REVIEW

**Comment by lipari** [comment]
Did you withhold the src/common/libutil/profiling.[ch] files for some reason?

**Comment by trws** [comment]
Thanks for the heads up don. Rebase was giving me fits and something must have gone wrong. I'll add them back in shortly.


________________________________

**Comment by trws** [comment]
@lipari: the files are in now.  I'm re-working how exactly caliper records information so it's a bit easier to parse, but the interface is unlikely to shift much.

One heads up though, don't compile the caliper library and flux with one of the dotkit gccs and try to use it on hype, the compute nodes don't have the appropriate libstdc++, so you'll get cryptic link errors...  @dongahn, is there someone to report this kind of thing to? I keep ending up fighting our compilers all over the place.

**Comment by dongahn** [comment]
@trws: I would start from John Gyllenhaal.

**Comment by trws** [comment]
@garlick, @grondo: This has had all TAU foolishness removed, been cleaned up somewhat, and rebased to current master.  There is something causing an event to be recorded multiple times, but it appears to be a caliper bug that I'm following up on (and it's easy to filter off events that have identical record, times, etc.).

Currently it tracks the rank, thread id, thread type (main or module thread), module name if present, message type and message topic for handle callbacks, and times all such message handlers.  If you want to give it a try, just use --enable-caliper and give the library and include paths for a copy of caliper, then run flux with `CALI_CONFIG_PROFILE=thread-trace` set.  There are a variety of extra options available through environment variables for caliper, but that will give pretty decent output.

The one other noticeable downside right now, that I would love some thoughts on if something comes to mind, is that flux/wreck processes under the brokers end up emitting profiles as well.  I haven't found a good way to turn off caliper at that boundary yet, but it should be reasonably possible with some environment filtering or similar.

**Comment by garlick** [comment]
Great, thanks!

It's easier to grok this PR as a single diff - might want to squash it down, and if there's any detritus remaining from tau (xmalloc? static htab decl. in wrexec), drop it.

Is the global state essential to communicating with caliper, or could it be moved to `struct dispatch` or maybe assigned to the `flux_t` handle with `flux_aux_set()`?

If we're making caliper the "native" trace format, can the indirection through profiling.h be dropped?

I will get caliper built and play with it a bit and probably then all will become clear.  Thanks again for getting this all going.

**Comment by garlick** [comment]
Is there a reason both `--with-caliper` and `--enable-caliper` are in configure.ac?

**Comment by trws** [comment]
Completely agree @garlick, I have every intention of squashing it before merge, but there are some things from the history that I may want to put in a future PR ( I had part of a compiler auto-instrument hook in there ) to pull out first.

The global state can be somewhere else, but all they are is named event types.  They should be common to the whole process, and there should only ever be one copy of those variables.   I have no issue with moving them if you prefer, but there should never be more than one of each (or we may get multiple attributes with the same name and type, which would be confusing).

I can certainly remove the header, but I wasn't sure whether to do it or not.  The header allows the common calls to be compiled out without too many guards, but it's no more than a convenience.

If they're both still there, it's by mistake, the `--enable-caliper` version is the real one.

Clearly this need further cleanup, so I'll get to that shortly and post an update.

**Comment by garlick** [comment]
No rush!  I've got caliper built on my desktop and I'll figure out how to get some trace output tomorrow.

**Comment by grondo** [comment]
> The one other noticeable downside right now, that I would love some thoughts on if something comes to mind, is that flux/wreck processes under the brokers end up emitting profiles as well. I haven't found a good way to turn off caliper at that boundary yet, but it should be reasonably possible with some environment filtering or similar.

If you unset caliper environment variables in the initial program shell does that fix a large part of the problem? For `wrexecd` or other flux components directly forked from the broker we will have to filter the environment as you say, or do we want to go ahead and keep the caliper data for these processes?

**Comment by grondo** [comment]
One more (probably very naive) question. In @tpatki's work with tracing in the scheduler we had the idea to start tracing with an event or other request, allowing tracing to be started/stopped dynamically. Would this ever be possible with caliper? (Doesn't need to be now, but dynamic tracing would be extremely handy in the future)

**Comment by trws** [comment]
Unsettling the end really doesn't do it because of they way the wreck processes are forked. This might be part of choosing when we do and don't want tracing to happen. It's definitely possible to do targeted tracing with caliper, but I haven't worked out the best way to manage that yet. It has built in support for some kinds of filtering and white/blacklisting, but we probably want our own event classes or regions to make that reasonably usable.


________________________________

**Comment by grondo** [comment]
> Unsettling the end really doesn't do it because of they way the wreck processes are forked. 

Sorry I'm being dense but I don't follow that, I'm not sure of how wreck processes are forked in any special way, but this would be easy to change if it helps here, and is good feedback for the replacement.

>  It's definitely possible to do targeted tracing with caliper, but I haven't worked out the best way to manage that yet. It has built in support for some kinds of filtering and white/blacklisting, but we probably want our own event classes or regions to make that reasonably usable.

Yeah, I was thinking the same thing, if we had our own filters, possibly dynamic like linux ftrace, that would be extremely handy. However, I looked at how ftrace accomplishes near zero overhead when tracing is disabled and it is suitably tricky that I don't think it is something we'd attempt.

I was hoping we'd be able to hide any complexity behind a `flux trace` command, but that might be well down the road...

**Comment by grondo** [comment]
> I have every intention of squashing it before merge, but there are some things from the history that I may want to put in a future PR ( I had part of a compiler auto-instrument hook in there ) to pull out first.

You could push a copy of this branch first
```
git push origin tau-tracing:tau-tracing-saved
```

Then go ahead and modify the PR branch as needed and `tau-tracing-saved` will always be available with your old commits.

**Comment by garlick** [comment]
Incidentally, d1932412a391a120a4777db07420fc5016592783 is a simplified version of your pr
* rebased on master
* dropped profiler.h
* moved `cali_id_t`'s to struct dispatch (according to [docs](http://software.llnl.gov/Caliper/api.html) `cali_create_attribute()` can be called multiple times and only one actual attribute will be created so pthread_once is maybe not needed?)
* dropped xrealloc change
* dropped lua debugging
* dropped wrexecd htab static

Didn't mean to gut your abstractions but I found this a bit easier to map directly to the caliper documentation for starting out.  Feel free to ignore.

**Comment by garlick** [comment]
I didn't get very far yesterday with some interruptions, but did have this thought (and started writing some code for it though not there yet):  it seems that caliper gets initialized and produces a zero length output file for an executable linked to it, even if no caliper calls are made.   To get control of this, it seems like caliper should only be linked to executables that require tracing, for starters the broker. 

To get trace calls added to the dispatcher without linking caliper to libflux.so, perhaps we could implement a trace registration interface that registers begin/end callbacks for specific topic strings or f`struct flux_match` specs?  Then the callbacks (back in the broker, say) would be the only bits required to link with caliper.

Any feelings about this approach?  Can we come up with a generic `flux_trace_f` callback signature that could work for begin/end callbacks in the different subsystems that we might like to trace?  I was thinking something as simple as this could work:
```C
typedef void (*flux_trace_f)(const char *name, void *arg);
```
(In the dispatcher I would construct `name` to be `message-type:topic-string`.

**Comment by grondo** [comment]
Could we also push a patch to caliper that opens output files dynamically, and only when there is data to write?

I like your callback idea. I'd really like to see the ability to have tracing enabled by default in the build with a `flux trace` frontend to enable/disable it at runtime, with message and function matching as you describe. Compile time tracing is useful, but runtime tracing will be very very useful, IMO. 

However, if I'm the only one asking for this feel free to ignore me.

**Comment by trws** [comment]
@grondo, as to the patch, it wouldn't be a bad thing to have a caliper plugin that does something more sensible for flux in general for sure.

The specific avoid opening until there's output patch that is something I imagine @daboehme would take if someone wants to make the adjustment.

As to using callbacks for tracing, I'm generally for having some of those mechanisms, especially since that's how compiler-instrumentation works at runtime it's good to have a way to tie into that. That said, it means we have to build rather more on a per-trace-target basis than we would otherwise, and every call into it will be a bit more expensive.

Either way, having a way to filter what does and doesn't get logged is going to be important, so the code for selecting a pattern or set of specs would be useful regardless.  I have an appointment set up with @daboehme to talk about how to use caliper with flux, and how to use the built-in filtering stuff, this Friday if anyone is interested in joining in.  I want to have a better handle on that before building too much abstraction that might be duplicative.  Speaking of, I think you're right @garlick that it's best to drop the profiling.h header, at least until we have a generic interface or something we want to use, with TAU gone there's no reason for it, so I'll probably merge your commit and add some of the adjustments I've been working on to that.

**Comment by garlick** [comment]
I'm going to be away this afternoon, so in case it's useful, my cruedly formed dispatch hook is working and available on my caliper branch [here](https://github.com/garlick/flux-core/tree/caliper), specifically

88df4dd2ad89490318c8d601a63aac176a9b9df9 broker: call dispatch trace hooks
59355e983c6d9b0226a50a772305a2726f679ef0 libflux/dispatch: add trace hooks to dispatch

**Comment by garlick** [comment]
Just a clarification: I think we get empty caliper logs only if we set the CALI_CONFIG_PROFILE environment variable.  If that's set (e.g. inherited by wrexecd) then we get a log file per process.

That seemed odd to me because wrexecd didn't call cali_* at all internally, but I could see the argument for it:  if you enable tracing and none of the trace points happen to be executed, that's a valid trace result.  So maybe not a bug?

**Comment by grondo** [comment]
That seems more reasonable to me.

**Comment by garlick** [comment]
dlopening libcaliper might be a way to enable Flux to be built with tracing that is enabled at runtime, and to avoiding getting trace files where they are not wanted.  There is not much in the [C API](https://github.com/LLNL/Caliper/blob/master/src/caliper/cali.h) to make accessing it via dlsym super incovenient.

**Comment by tpatki** [comment]
Should I move the flux-sched tracing to caliper as well? @dongahn wanted a simpler tracing mechanism (what I was working on earlier) but caliper was Martin/my original suggestion too. What do you all think?

**Comment by lipari** [comment]
> Should I move the flux-sched tracing to caliper as well?

Sounds good to me.

**Comment by grondo** [comment]
I like @garlick's idea -- it would be interesting to see if there is a measurable performance overhead using this approach. @tpatki, it would be cool if sched tracing uses a similar scheme @garlick is working toward, then all flux tracing would come out in a similar format to the same files. That would be cool.

We still might want some scheme for filtering which tracepoints get enabled

**Comment by garlick** [comment]
> Should I move the flux-sched tracing to caliper as well?

Well, I think we are headed towards a generic trace framework provided by core, but it wouldn't hurt for you to switch over to caliper if you don't mind switching again once we have sorted out the generic hooks.

At least that's what I was thinking.  @trws should chime in with hits thoughts - I hope he doesn't feel I've hijacked his PR :-)

**Comment by tpatki** [comment]
@trws and I discussed this. (Sorry for hijacking the conversation). 

It seems like it might be a good idea to have a common way of tracing core modules and sched (through an event/message interface along with caliper). It might be easier to add caliper extensions if needed. Also, we need a way to specify the depth/granularity of profiling, esp with sched where we don't want a flat profile. Another thing that @dongahn and I had discussed is to have a parameter that determines when to write aggregated info to a file, and this is easier to do with Caliper from what I gather.

**Comment by trws** [comment]
@garlick, I think you're right that that's the direction, but the trace hooks you put together so far are relatively message-centric and build on caliper internally.  @tpatki has a similar mechanism for modules, something I thought we might even want to pull into the base module loader somewhat like the comm stats stuff, that lets users register their own tracing functions while calling through a standardized interface so it can control whether trace points are actually active or not dynamically.  The caliper part is more just the actual tracing code.  By the same token, if we want to set an attribute somewhere to be used by the trace points, is the plan to wrap those as well or just use the caliper interface? Given the last few comments I'm less sure we're all on the same page than I was.

**Comment by garlick** [comment]
Sorry for the noise - I concluded I was heading in the wrong direction yesterday, and didn't really have time to figure it out.  I felt like the logical evolution of those tracepoint callbacks I introduced was going to be to reinvent caliper, which of course we shouldn't do. Please ignore them.

The unfinished thought I had yesterday was would it make any sense to create a set of thin wrappers for the caliper C API so that libcaliper could be dlopened when tracing is enabled at runtime?  Then caliper wrappers could be used at tracepoints, and the broker or other traceable programs could be wired with a message handler or options that enable/disable tracing and manipulates CALI_  environement variables.  The whole point of this would be to leverage caliper but allow it to be compiled in all the time with near zero overhead (libcaliper constructor wouldn't run unless requested).

It sounds like you and @tpatki had a good discussion and I am comfortable with you two driving this plan (don't worry if I didn't seem like I was on the same page - it's because I was just fumbling around!)

**Comment by garlick** [comment]
Possibly interesting LLNL trace visualization project:  [ravel](https://github.com/LLNL/rave?   It is based on Open Trace Format which I've heard mentioned but am not familiar with and do not know how it relates to Caliper's trace format.

**Comment by trws** [comment]
Caliper is orthogonal to OTF, but contains all the information such that 
if someone writes a converter for it, it should work.  Several folks 
related to the project have extensive experience with OTF, so it's 
probably just a matter of time before that happens.

**Comment by garlick** [comment]
I restarted the travis gcc build which had failed on the jstat test (known issue #512).

**Comment by trws** [comment]
Thanks @garlick!

**Comment by garlick** [comment]
I'm running into this with caliper 1.4.0,  [the lastest github release](https://github.com/LLNL/Caliper/releases).   I'll try building caliper from the git master.
```
  CC       libflux_la-handle.lo
handle.c: In function ‘msg_snapshot’:
handle.c:508:5: error: too many arguments to function ‘cali_push_snapshot’
     cali_push_snapshot (
     ^
In file included from handle.c:37:0:
/usr/local/include/caliper/cali.h:114:1: note: declared here
 cali_push_snapshot(int scope);
 ^
Makefile:985: recipe for target 'libflux_la-handle.lo' failed
make[3]: *** [libflux_la-handle.lo] Error 1
make[3]: Leaving directory '/home/garlick/proj/flux-core/src/common/libflux'
Makefile:649: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/home/garlick/proj/flux-core/src/common'
Makefile:460: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/garlick/proj/flux-core/src'
Makefile:534: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

**Comment by tpatki** [comment]
Anyone tried building this branch with flux-sched? Some tests are failing with: 

pmi-mapping.lua: rexecd_init: ...amework/flux-install/etc/wreck/lua.d/pmi-mapping.lua:8: attempt to perform arithmetic on local 'nnodes' (a nil value)

I'm trying to dig deeper now but have an all-afternoon meeting with Sandia so thought I'd at least let you all know.

**Comment by garlick** [comment]
@tpatki - this branch is a little behind the current master and in fact doesn't include the pmi-mapping.lua script.  I'm not sure where you might be picking it up, but I would suggest cleaning everything and rebuilding.

**Comment by garlick** [comment]
@trws - I think we knew this, but `make check` fails here
```
make[4]: Entering directory '/home/garlick/proj/flux-core/src/common/libflux'
  CC       test/test_module_t-module.o
  CCLD     test_module.t
../../../src/common/libflux/.libs/libflux.a(libflux_la-rpc.o): In function `rpc_get':
/home/garlick/proj/flux-core/src/common/libflux/rpc.c:233: undefined reference to `cali_begin_string_byname'
/home/garlick/proj/flux-core/src/common/libflux/rpc.c:238: undefined reference to `cali_end_byname'
```

Would it make sense in this case to append `-lcaliper` to LIBS so that adding tracing to any point in the code base when it's enabled won't require a Makefile.am tweak?

**Comment by garlick** [comment]
Also we're defining PROFILING in the configure.ac but not using it at all.

**Comment by tpatki** [comment]
@garlick: I already tried that yest, but let me try again.

**Comment by garlick** [comment]
In `handle.c`, some suggestions to make the flow of the handle code easier to follow for someone jumping in there for the first time
* add a `struct profiling_context` or similar to group the profiling-related elements that were added to directly `flux_handle_struct` 
* move the `cali_create_attribute()` calls that were added to flux_open to a `profiling_context_init()` function
* rename `msg_snapshot()` should probably be renamed to e.g. `profiling_msg_snapshot()` .
* ifdef out the calls to these functions (rather than their main body) so spelunker immediately knows code does not have side effects if tracing is not enabled.

Same comments apply (to a lesser degree) in rpc.c and dispatch.c.

The trace data being snapshotted seems right but i'd like to run with tracing and see what the traces look like.  Have to dig a bit to remember how to do that.  Which incidentally, it may be a good idea to provide a tracing cheat sheet somewhere.   Maybe a `flux-tracing(7)` man page?

Please follow the RFC 7 conventions for spacing function names and params and breaking lines at 80 cols.

**Comment by garlick** [comment]
I did notice cali-query produced this warning on stderr when I post processed a trace file:
```
$ CALI_CONFIG_PROFILE=thread-trace ./flux start /bin/true
$ cali-query *.cali >cali.out
vi== CALIPER: Invalid node record format: __rec=node,attr=18446744073709551615,data=,id=118
== CALIPER: Invalid node record format: __rec=node,attr=18446744073709551615,data=,id=119,parent=118
== CALIPER: Invalid node record format: __rec=node,attr=18446744073709551615,data=,id=120,parent=119
```

Any chance of having your python script to turn query output into something more meaningful to humans?   Perhaps it could be added to src/test/ as part of this PR?  I'm leaving for a doctor's apt but I'll try to get back to this in the afternoon.

**Comment by grondo** [comment]
> @tpatki - this branch is a little behind the current master and in fact doesn't include the pmi-mapping.lua script. I'm not sure where you might be picking it up, but I would suggest cleaning everything and rebuilding.
> > @garlick: I already tried that yest, but let me try again.

If your copy has `pmi-mapping.lua` script then `make clean` won't be able to clean it  on this branch since Makefiles don't know about it yet.

I'd start with a fresh clone, (carefully) use `git clean`, or just `rm -f src/modules/wreck/lua.d/pmi-mapping.lua`

**Comment by trws** [comment]
Rebased to master, working my way down the comments right now, so there 
may be churn for a bit.

**Comment by trws** [comment]
Ok, I think I've addressed the majority of the comments here.  There is still something odd, in that we get "caliper initialized" messages even when running unit tests with no caliper environment variables set, I'm not sure why this is but it needs to go away.  Aside from that, here's a snippet of python that one can use to do some parsing on the result:

```python
#!/usr/bin/env python
import sys
import os
import re
import subprocess as sp
import pandas as pd

reader = sp.Popen([os.environ['HOME'] + "/build/Caliper/install/bin/cali-query", "-e"] + sys.argv[1:], stdout=sp.PIPE)

events = []
for l in reader.stdout:
    d = {}
    events.append(d)
    for kv in l.strip().split(','):
        kvs = kv.split('=')
        key = kvs[0]
        try:
            value = int(kvs[1])
        except:
            value = kvs[1]
        d[key] = value

df = pd.DataFrame(events)

print df[df['flux.message.topic'] == 'kvs.put'].drop_duplicates().sort('time.offset')
```

You'll need pandas to use the dataframe, and clearly this isn't optimal, but there is an issue wth the python reader so the run query and split method is a temporary workaround.  The print at the end filters the output to show only those events that are based on messages sent with the "kvs.put" topic, de-duplicated (because of the forking issue) then sorted by the time.  To have time.offset, you'll need to add the CALI_TIMER_OFFSET=true variable if I recall correctly.

Evidently there are also travis issues, will look into that shortly.

**Comment by trws** [comment]
I must have somehow caused this travis issue, but I have to admit I'm at a loss as to *how* precisely I caused it.  @grondo, @garlick, by any chance could you take a look at the travis log?  It seems to be saying that somehow pkg-config doesn't exist for the configure script right now, but I have no clue how I could have caused that with what I added.

**Comment by trws** [comment]
Especially since setting PKG_CONFIG=/usr/bin/pkg-config seems to "fix" it, I'm at a loss.

**Comment by grondo** [comment]
@trws, I'll take a look at Travis-CI issue

**Comment by grondo** [comment]
> @trws, I'll take a look at Travis-CI issue

The problem reproduced on hype (TOSS2 system). I was able to workaround the problem by moving the `LIBCALIPER` config *after*  `X_AC_ZEROMQ`. It would be nice to understand this, so I'm still poking at it, but moving the caliper stuff in `configure.ac` might get you moving forward for now.h

**Comment by trws** [comment]
Thanks @grondo, I'll make that change for now at least.

**Comment by grondo** [comment]
Oh geez, this seems to be the problem from [Autotools Mythbuster, S3.4](https://autotools.io/pkgconfig/pkg_check_modules.html):

> Sometimes, you're supposed to check for given modules only under some conditions; it's a trivial setup, but it's one of the most common mistakes. The pkg-config command is discovered through a separate macro, PKG_PROG_PKG_CONFIG that takes care of identifying the presence (and version) of pkg-config itself. This macro is called through AC_REQUIRE so that is expanded before PKG_CHECK_MODULES.
>
> If you have the first call to PKG_CHECK_MODULES inside a bash conditional block, the expansion of PKG_PROG_PKG_CONFIG will also be conditional, so the following code will fail to work, when the condition is false:

> ```
> AC_ARG_WITH([gtk], AS_HELP_STRING([--with-gtk], [Build with the GTK+ interface]))
> 
> if test "x$with_gtk" = "xyes"; then
>    PKG_CHECK_MODULES([GTK], [gtk+-2.0])
> fi
>
> PKG_CHECK_MODULES([GLIB], [glib-2.0])
> ```
> Since the check for GTK+ will not work by default, you'll receive the following error if you try to execute configure without any argument:
> 
> ```
> checking for GLIB... no
> configure: error: in `/path':
> configure: error: The pkg-config script could not be found or is too old.  Make sure it
> is in your PATH or set the PKG_CONFIG environment variable to the full
> path to pkg-config.
> ```
> 
> You can solve this problem in two ways; you can either explicitly call PKG_PROG_PKG_CONFIG outside of any conditional, forcing checking for pkg-config as soon as possible; or you can rewrite your conditionals to use the proper syntax, as discussed in Section 1, “M4sh”.
> 
> The proper code written in M4sh for the above logic is the following:
>```
> AC_ARG_WITH([gtk], AS_HELP_STRING([--with-gtk], [Build with the GTK+ interface]))
> 
> AS_IF([test "x$with_gtk" = "xyes"], [
>    PKG_CHECK_MODULES([GTK], [gtk+-2.0])
> ])
>
> PKG_CHECK_MODULES([GLIB], [glib-2.0])
> ```

**Comment by grondo** [comment]
Given the above I say we add a commit to this PR that calls `PKG_PROG_PKG_CONFIG` early in `./configure.ac` (with a comment as to why), well before the first use of `PKG_CHECK_MODULES`

**Comment by trws** [comment]
Makes perfect sense to me.  I'll add one right now.  Thanks for tracking that down @grondo!

**Comment by daboehme** [comment]
Just my 2ct:
@trws The "caliper initialized" message is always displayed when caliper is linked in and caliper calls are made, just to let you know it's there. You can shut it up by setting CALI_LOG_VERBOSITY=0.
@garlick The "invalid node record" messages you saw earlier indicate that something went seriously wrong. I would like to investigate this if possible.

**Comment by garlick** [comment]
> The "invalid node record" messages you saw earlier indicate that something went seriously wrong. I would like to investigate this if possible.

@daboehme  - thanks for following up.  I will run the same test a few more times and see if it reproduces reliably, and can also get you the trace file to look at.  Anything else you would like me to try?

**Comment by trws** [comment]
I actually have at least one other trace set that gives me that error, or one much like it, for each process's file. The test case is simple enough I could forward it along if that would be helpful.


________________________________

**Comment by garlick** [comment]
Here is the trace file that I was commenting on:
[160707-105957_809_5mXXnIvgmyzI.cali.txt](https://github.com/flux-framework/flux-core/files/354581/160707-105957_809_5mXXnIvgmyzI.cali.txt)

**Comment by garlick** [comment]
I ran that the test above three more times and got that error from cali-query each time.

**Comment by daboehme** [comment]
@garlick Thanks for the trace file.
I think the only way the invalid nodes could get in there is through the `cali_push_snapshot` call. Now Caliper shouldn't do that even if some arguments are invalid, so that's a bug in caliper. @trws, can you verify that all the attributes and values you put in the `trigger_info` lists are correctly initialized? If so, is there something missing in the expanded output?

**Comment by garlick** [comment]
So I think we can get this merged once we've cleared the following hurdles:
* Fix the "invalid node record" problem
* (as discussed in Flux meeting) centralize the caliper build options (append to LIBS, and whatever CPPFLAGS variable is appropriate?)
* Drop incremental development from commits and rebase on current master.

**Comment by trws** [comment]
@daboehme, I've put in some time looking for the invalid node record issue.  It's definitely related to using the C interface snapshot function, in that I get one invalid node record for each element I place in the snapshot, once per run.  The data being snapshotted appears to be correct, but I still get the invalid node with no data and a UINT64_MAX attr.  Any thoughts?

**Comment by trws** [comment]
Ok, I think this last push addresses the remaining issues.  The invalid nodes were being created because the profiling call was being invoked inside of connector_init by messaging happening after it initialized its part of the handle but before letting the handle finish initializing itself.  Also rebased and squashed down.

Review @garlick, @grondo?

**Comment by grondo** [comment]
Thanks! I'll take a look asap, hopefully by early tomorrow.

**Comment by grondo** [comment]
@trws, I sanity checked this PR and it all looks pretty good!

Some minor comments:
 1. Did you address the second of @garlick's comments above?

> (as discussed in Flux meeting) centralize the caliper build options (append to LIBS, and whatever CPPFLAGS variable is appropriate?)

 2. Probably want to remove wording the commit message that references prior versions of the PR that won't exist in permanent git history. Possibly rewrite commit message explaining just what  is in the commit?

 3. One of the `t2000-wreck.t` tests fail without setting `CALI_LOG_VERBOSITY=0` (due to the unexpected `CALIPER` messages on stderr). This isn't a critical thing to get fixed, but I wonder if @daboehme would consider making `0` the default log verbosity, or we could simulate that in `flux` by something like `if (!getenv ("CALI_LOG_VERBOSITY")) setenv ("CALI_LOG_VERBOSITY", "0", 0)` before the first Caliper call. (Unless caliper is initialized in a library constructor?)

**Comment by trws** [comment]
I did, or at least thought I did, address @garlick's comment about the flags, and rewording the commit message is no problem.  As to the logging, there's actually a bit of a choice there.  I can actually build flux so that it builds against the stubs library, then use LD_PRELOAD to load in caliper when we actually want results.  Alternately, we can control the logging and other things, but having it use the stubs by default should fix all of the testing issues, and make a build with caliper built in run faster when caliper's real implementation isn't loaded.  What do you think would make more sense there? Either is equally easy honestly.

**Comment by grondo** [comment]
Ok, thanks, sorry I missed your changes related to @garlick's comment!

For the Caliper logging issue, I guess it is a matter of taste, but my personal preference is that libraries do not log to stderr by default for informational messages like "hello" and "goodbye" ;-)

Even when I'm using Caliper purposely to trace a 100 broker job, I'm probably not going to want to see 100 "Initialized" messages on stderr I guess is my point.

The dlopen idea sounds great to me!  Perhaps as a follow on to this PR?  If you fix up the commit message I'd be willing to merge this PR as is so that you aren't required to continually rebase on master (and it will be nice to have some kind of canned tracing support in the next flux-core tag)

Thanks!

**Comment by trws** [comment]
I completely agree with that sentiment for what it's worth... @daboehme, any chance of a default change on that at some point? For now, we can manipulate the environment variables or something, but for a library we want to keep loaded I'm not sure it's a good thing.

Let me take a minute to see if I can do the switch to stubs easily, failing that we'll make that part of another pr.

**Comment by grondo** [comment]
I poked at this Friday and over the weekend a bit. The stubs seem to work fine, but I do have to make the following change to `./configure.ac` to build with libcaliper-stubs installed in a non-standard location: 
(missing `caliper` arg to `pkg-config`)

```diff
--- a/configure.ac
+++ b/configure.ac
@@ -124,7 +124,8 @@ AC_ARG_ENABLE(caliper,
 if test "$enable_caliper" = "yes"; then
     PKG_CHECK_MODULES([CALIPER], [caliper], [], [])
     CFLAGS="${CFLAGS} ${CALIPER_CFLAGS} "
-    LIBS="${LIBS} $(pkg-config --libs-only-L ) -lcaliper-stub -lrt "
+    # Do not use CALIPER_LIBS, only link to libcaliper-stub
+    LIBS="${LIBS} $(pkg-config --libs-only-L caliper) -lcaliper-stub -lrt "
     AC_DEFINE([HAVE_CALIPER], [1], [Define if you have libcaliper])
 fi
```

 was also thinking, would it be a total kludge to add a `--profile` option to `flux-start` that sets `LD_PRELOAD=libcaliper.so`?

Alternately, if we pull in the stubs library ourselves, we could perhaps redirect symbols at runtime with `dlopen()` instead of preloading them, and allow tracing to be enabled and eventually disabled at runtime? (I'm not sure of the performance impact, however, and this could be future work)

**Comment by garlick** [comment]
> would it be a total kludge to add a `--profile` option to `flux-start`

I wouldn't object to that.

**Comment by grondo** [comment]
Something like the below seems to work -- the option only appears if flux-core is built with `--enable-caliper` -- An alternative would be to always build with the option, but return an error if the option is used when built without caliper support.

We could also set some other `CALIPER_*` environment variables under `--profile` (without overwrite of course), but I don't know of a reasonable default -- or perhaps the Caliper configuration profile could be given on the cmdline, e.g. `--profile=thread-trace`?

Hmm, maybe `--profile` should be named `--caliper-profile`

```diff
diff --git a/src/cmd/flux-start.c b/src/cmd/flux-start.c
index c40d46f..fa3a755 100644
--- a/src/cmd/flux-start.c
+++ b/src/cmd/flux-start.c
@@ -96,6 +96,10 @@ static struct optparse_option opts[] = {
       .usage = "Add comma-separated broker options, e.g. \"-o,-q\"", },
     { .name = "killer-timeout",.key = 'k', .has_arg = 1, .arginfo = "SECONDS",
       .usage = "After a broker exits, kill other brokers after SECONDS", },
+#if HAVE_CALIPER
+    { .name = "profile",.key = 'P', .has_arg = 0,
+      .usage = "Run brokers with Caliper profiling enabled", },
+#endif
     OPTPARSE_TABLE_END,
 };
 
@@ -387,6 +391,22 @@ struct client *client_create (struct context *ctx, int rank
     subprocess_setenvf (cli->p, "PMI_FD", 1, "%d", client_fd);
     subprocess_setenvf (cli->p, "PMI_RANK", 1, "%d", rank);
     subprocess_setenvf (cli->p, "PMI_SIZE", 1, "%d", ctx->size);
+
+#if HAVE_CALIPER
+    /*
+     *  If --profile was used, set or append libcaliper.so in LD_PRELOAD
+     *   to subprocess environment, swapping stub symbols for the actual
+     *   libcaliper symbols.
+     */
+    if (optparse_hasopt (ctx->opts, "profile")) {
+        const char *pl = getenv ("LD_PRELOAD");
+        subprocess_setenvf (cli->p, "LD_PRELOAD", 1, "%s%s%s",
+            pl ? pl : "",
+            pl ? " ": "",
+            "libcaliper.so");
+    }
+#endif
+
     return cli;
 fail:
     client_destroy (cli);

```

**Comment by garlick** [comment]
I like where you ended up: `--caliper-profile[=type]` and option present but returning helpful error if built without caliper.  (No reason to give them "bad option" and make them sleuth out the reason).

**Comment by grondo** [comment]
Ok, something like this? 
(Though in general I do not have a problem with cmdline options that come and go with optional code -- we will likely eventually have that when plugins are able to provide new options to utilities in the future -- it does seem valid in this case I think) 

```
$ src/cmd/flux start --help
Usage: flux-start [OPTIONS] command ...
  -X, --noexec           Don't execute (useful with -v, --verbose)
  -h, --help             Display this message.
  -k, --killer-timeout=SECONDS
                         After a broker exits, kill other brokers after SECONDS
  -o, --broker-opts=OPTS Add comma-separated broker options, e.g. "-o,-q"
  -s, --size=N           Set number of ranks in new instance
  -v, --verbose          Be annoyingly informative
      --caliper-profile=PROFILE
                         Enable profiling in brokers using Caliper configuration
                         profile named `PROFILE'
```

```
$ src/cmd/flux start /bin/true

$ src/cmd/flux start --caliper-profile=thread-trace /bin/true
== CALIPER: Registered event trigger service
== CALIPER: Registered pthread service
== CALIPER: Registered recorder service
== CALIPER: Registered timestamp service
== CALIPER: Registered trace service
== CALIPER: Initialized
2016-08-01T23:45:45.596233Z broker.err[0]: rc1: == CALIPER: Initialized
2016-08-01T23:45:45.665101Z broker.err[0]: rc1: == CALIPER: Initialized
2016-08-01T23:45:45.747175Z broker.err[0]: rc1: == CALIPER: Initialized
2016-08-01T23:45:45.818867Z broker.err[0]: rc1: == CALIPER: Initialized
2016-08-01T23:45:45.872114Z broker.err[0]: rc1: == CALIPER: == CALIPER: InitializedInitialized
2016-08-01T23:45:45.964693Z broker.err[0]: rc1: == CALIPER: Initialized
2016-08-01T23:45:46.055275Z broker.err[0]: rc1: == CALIPER: Initialized
2016-08-01T23:45:46.167584Z broker.err[0]: rc3: == CALIPER: Initialized
2016-08-01T23:45:46.224110Z broker.err[0]: rc3: == CALIPER: Initialized
2016-08-01T23:45:46.278749Z broker.err[0]: rc3: == CALIPER: Initialized
2016-08-01T23:45:46.334260Z broker.err[0]: rc3: == CALIPER: Initialized
2016-08-01T23:45:46.389870Z broker.err[0]: rc3: == CALIPER: Initialized
2016-08-01T23:45:46.445450Z broker.err[0]: rc3: == CALIPER: Initialized
2016-08-01T23:45:46.500377Z broker.err[0]: rc3: == CALIPER: Initialized
== CALIPER: Trace: Flushed 622 snapshots.
== CALIPER: Wrote 975 records.
```

Built without `--enable-caliper`
```
$ src/cmd/flux start --caliper-profile=thread-trace /bin/true
flux-start: Error: --caliper-profile used but no Caliper support found
```

**Comment by garlick** [comment]
That looks great to me!

**Comment by garlick** [comment]
Closing as we've merged #741.
